### PR TITLE
Add ability to parse Annex B stream in FU-A

### LINKED ISSUE #68: Exiting because of bad sps
Hi, I tried out your crate to check if it works with my outdoor camera. It exits out because of `bad sps`.

Is this a problem with how the camera implements RTSP?

Thank you

## Output
```bash
I20221015 14:37:53.671 main client::mp4] Using h264 video stream
I20221015 14:37:53.686 main client::mp4] No suitable audio stream found
E20221015 14:37:55.589 main client] Fatal: bad sps

conn: 192.168.10.15:53286(me)->192.168.10.10:554@2022-10-15T14:37:53
stream: TCP, interleaved channel ids 0-1
ssrc: 00000000
seq: 00000068
pkt: 150505@2022-10-15T14:37:54
```

### PR DESCRIPTION
**Pull request by thatdevsherry**
# Summary
Add ability to break apart an Annex B stream sent in a FU-A. Fixes #68 

# Details
My V380 cam would send a FU-A after establishing RTSP connection. The FU-A was not conformant to spec.

```
FU-A (start) => [ sps header - sps - boundary - pps header - pps - boundary - idr header - idr ] # Annex B stream
FU-A (...) => [ sps header - idr ]
FU-A (end) => [ sps header - idr ]
```

Notice how all the frags have the same header (as they should be), but the **start** has an Annex B stream, meaning the last NAL picked from that packet is an **IDR**. This means the last NAL saved from first packet will be an **IDR**, but the next fragment will have... the header for **SPS**, but data for **IDR**, which is wrong.

It appears that the camera only does this for FU-A that has SPS & PPS. FU-As & single NAL units for other NAL types are conformant to spec.

I have only modified this logic for the FU-A flow. We can however use the `start_rtp_nal`, `append_rtp_nal` and `end_rtp_nal` to handle all NAL types.

### Camera details
Name: V380 (It's a generic V380 outdoor camera)
Firmware: `HwV380E31_WF8_PTZ_WIFI_20201218` (I had asked them for a firmware update file to enable RTSP support)

### CODE REVIEW

File: src/codec/h264.rs [hunk 1]
```diff
@@ -1544,4 +1711,163 @@ mod tests {
...
+                loss: 0,
+                mark: false,
+                payload_type: 0,
+            }
+            .build(*b"\x3c\x07a wild sps appeared")
+            .unwrap(),
+        );
+        assert!(push_result.is_err());
+    }
 }
```
Comment (scottlamb) [inline src/codec/h264.rs:1873]: I'd like to have tests of `NalParser` by itself. That'd help explain its responsibility and isolate bugs to one struct or the other.

One test in particular seems valuable: two smaller calls to `append_rtp_data` should produce the same result as a longer one, even if they're in the middle of a run of zeros. This would help us ensure if the camera makes a long Annex B sequence and then breaks it apart into FU-As, it will parse correctly regardless of whether the FU-A boundaries line up nicely with the Annex B boundaries or not. I think we should be able to do something roughly like this and get it to pass:

```rust
#[test]
fn test_small_appends() {
    init_logging();
    let full_input = b"first nal\x00\x00\x01\x67second nal starts\x00\x00second nal continues";

    // Parse all at once.
    let mut all_at_once_p = NalParser::new();
    all_at_once_p.start_rtp_nal(...).unwrap();
    all_at_once_p.append_rtp_data(Bytes::from_static(full_input)).unwrap();
    all_at_once_p.end_rtp_nal(...).unwrap();
    assert_eq!(all_at_once_p.nals.size(), 2);
    // ...verify each nal looks right...

    // Parse in two pieces, at every possible boundary. Should get the same result.
    for i in 1..full_input.len() {
        let mut split_p = NalParser::new();
        split_p.start_rtp_nal(...).unwrap();
        let parts = full_input.split_at(i);
        split_p.append_rtp_data(Bytes::from_static(parts[0])).unwrap();
        split_p.append_rtp_data(Bytes::from_static(parts[1])).unwrap();
        split_p.end_rtp_nal(...).unwrap();
        assert_eq!(&all_at_once_p, &split_p); // add an `Eq` impl for this, or in some way test they get the same result
    }

    // Parse a byte at a time. Should get the same result.
    let mut byte_at_a_time_p = NalParser::new();
    byte_at_a_time_p.start_rtp_nal(...).unwrap();
    for i in full_input.len() {
        byte_at_a_time_p.append_rtp_data(Bytes::from_static(&parts[i])).unwrap();
    }
    byte_at_a_time_p.end_rtp_nal(...).unwrap();
    assert_eq!(&all_at_once_p, &byte_at_a_time_p); // likewise.
}
```

It may be tricky to get it to pass. In particular, if an `append_rtp_data` call ends right after a zero, I think it needs to put it in the buffer, and then a following call to `append_rtp_data` needs to take away that previous zero if it turned out to be part of a boundary! And likewise if there are two zeros (maybe in separate pieces) and then something other than a 0, 1, 2, or 3. That's not straightforward, so I'd find it really helpful to test this without the added complexity of bringing `Depacketizer`'s own responsibilities into it.

File: src/codec/h264.rs [hunk 2]
```diff
@@ -36,13 +36,182 @@ pub(crate) struct Depacketizer {
...
+                if self.seen_second_zero_at.is_none() {
+                    self.seen_second_zero_at = Some(idx);
+                    continue;
+                }
+
+                if self.seen_second_zero_at.unwrap_or(0) - self.seen_one_zero_at.unwrap_or(0) == 1
+                    && idx - self.seen_second_zero_at.unwrap_or(0) == 1
+                {
+                    debug!("Found boundary, idx range: {} - {}", idx - 2, idx);
+                    // we found a boundary, let NalParser know that it should now keep adding
```
Comment (scottlamb) [inline src/codec/h264.rs:111]: Nit: this is super picky of me, but could you match the style of other comments by having full English sentences? E.g. consistently starting with capital and ending with period.

File: src/codec/h264.rs [hunk 3]
```diff
@@ -36,13 +36,182 @@ pub(crate) struct Depacketizer {
...
+                        // to skip the packet type and NAL header
+                        nal = data.slice(nal_start_idx..nal_end_idx);
+                        nal_len = nal.len();
+                    } else {
+                        // ignore the first two bytes when saving NAL (Packet type header & NAL header)
+                        nal = data.slice(nal_start_idx + 2..nal_end_idx);
+                        nal_len = nal.len();
+                    }
+                    let pieces = self.add_piece(nal)?;
+                    let last_nal_saved = self.nals.last_mut().expect("nals empty while in fu-a");
```
Comment (scottlamb) [inline src/codec/h264.rs:131]: It took me a really long time to figure this out, but Rust's `expect` output is confusing if you word the message in terms of what's *wrong*. It's better if you word it in terms of what *should* always be true if our code is correct. Some helpful soul added an explanation here:

https://doc.rust-lang.org/std/error/index.html#common-message-styles

In this case, I might say `nals should be non-empty because start_rtp_nal should be called before append_rtp_nal`.

**Review by scottlamb — CHANGES_REQUESTED** [review]
I added a bunch of comments, but this is a great first pass! Thanks for digging in; it's really nice to get contributions like this.

**Comment by thatdevsherry** [comment]
> I added a bunch of comments, but this is a great first pass! Thanks for digging in; it's really nice to get contributions like this.

Thanks for taking the time out to review. I'll take a look at the comments and update the PR.

File: src/codec/h264.rs [hunk 4]
```diff
@@ -36,13 +36,182 @@ pub(crate) struct Depacketizer {
...
+struct NalParser {
     /// In state `PreMark`, pieces of NALs, excluding their header bytes.
     /// Kept around (empty) in other states to re-use the backing allocation.
     pieces: Vec<Bytes>,
 
     /// In state `PreMark`, an entry for each NAL.
     /// Kept around (empty) in other states to re-use the backing allocation.
     nals: Vec<Nal>,
+
+    // If true, retina does not panic when the header of a FU-A frag does not match
```
[the remark this answers is not in the document]
Comment (thatdevsherry) [reply src/codec/h264.rs:55]: Reg. the third point, could you please elaborate a bit on it?

From what I understand, in the FU-A code, we do check if FU-A headers match the starting FU-A header, and it exits if it doesn't match. We want to make it so that it ignores this inconsistency in case of Annex B stream.

How can we utilise `AccessUnit::in_fu_a` if we convert it from `bool` to `Option<NalHeader>`?

Is it that, in case of Annex B stream, we can then check the `AccessUnit::in_fu_a` header, and just ignore it inside the Annex B code block, rather than having an additional bool check (`NalParser::ignore_inconsistent_fu_a_headers`) that toggles the main header checking code?

If so, where would we move the header checking code so that it would exit on normal FU-A, but also not run for Annex B w/o needing to check for any other bool? It might have to go down into `start_rtp_nal` etc where it is closer to the Annex B code?

[the remark this answers is not in the document]
Comment (scottlamb) [reply src/codec/h264.rs:55]: I mean let's say there's a FU-A start packet for a SPS, and inside it there's an Annex B separator and the beginning of a PPS NAL. The next FU-A packet says it's continuing the SPS.

* without your `ignore_inconsistent_fu_a_headers` change, Retina fails, because it's expecting the FU-A to be continuing the latest NAL, which is a PPS.
* what I'm suggesting is using `in_fu_a = Some(NalHeader(0x67))` to mean we're expecting the FU-A to say it's continuing the SPS, rather than look at what the latest NAL we've decoded is.

and this would be all in `Depacketizer`; `NalParser` wouldn't need to know about single-packet NALs vs FU-A NALs vs STAP-A NALs.

[the remark this answers is not in the document]
Comment (thatdevsherry) [reply src/codec/h264.rs:55]: Oh my bad. I forgot that matching the FU-A header with last NAL **is** the original behaviour. Yeah, an `Option<NalHeader>` would be nice. I'll add it in upcoming PR commits.

File: src/codec/h264.rs (hunk 2 above)
Comment (thatdevsherry) [reply src/codec/h264.rs:106]: Updated in [e940ab4](https://github.com/scottlamb/retina/pull/71/commits/e940ab4a280e1930c891bf1c50a882557d5078b5)

File: src/codec/h264.rs (hunk 3 above)
Comment (thatdevsherry) [reply src/codec/h264.rs]: Thank you for sharing the link.

Updated in [d21a69a](https://github.com/scottlamb/retina/pull/71/commits/d21a69a7893a268498015f740e2460314c1d5594)

File: src/codec/h264.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (thatdevsherry) [reply src/codec/h264.rs]: I've updated `AccessUnit::in_fu_a` to be an `Option<NalHeader>` in [9cde15d](https://github.com/scottlamb/retina/pull/71/commits/9cde15df0e3bfdd2e049e1a1376b1841132d63bd).

Please let me know what you think of it.

File: src/codec/h264.rs [hunk 5]
```diff
@@ -36,13 +36,182 @@ pub(crate) struct Depacketizer {
 
     parameters: Option<InternalParameters>,
 
+    /// Holds NALs from RTP packets.
+    nal_parser: NalParser,
+}
+
+/// Contains logic that converts RTP payload into NALs.
```
[the remark this answers is not in the document]
Comment (thatdevsherry) [reply src/codec/h264.rs:43]: Should single NAL units only be handled by a single function e.g. `start_rtp_nal`? Even if it has an Annex B, `start_rtp_nal` will handle it.

```
# FU-A

FU-A (start) => start_rtp_nal # break Annex B
FU-A (...) => append_rtp_nal
FU-A (end) => end_rtp_nal

# Single NAL unit

Packet => start_rtp_nal # break Annex B
```

Also, since you mentioned that we might get cases where Annex B boundaries don't line up inside a single FU-A, and the boundary can get cut off b/w packets, we'd need to run Annex B reader in the `append_rtp_nal` as well as `end_rtp_nal` as well.
```
FU-A (start) => start_rtp_nal # break Annex B
FU-A (...) => append_rtp_nal  # break Annex B
FU-A (end) => end_rtp_nal     # break Annex B
```

Do I understand this correctly?

[the remark this answers is not in the document]
Comment (thatdevsherry) [reply src/codec/h264.rs]: Ah yes, I got a glimpse of this when looking at the sample test case you quoted [here](https://github.com/scottlamb/retina/pull/71#discussion_r1008889113) 😄 . This is really nice, we can leave Annex B parsing to `append_rtp_nal`, and it's easier to test.

The responsibilities of these functions would then be:
```
start_rtp_nal => create a new `Nal` only, no processing whatsoever
append_rtp_nal => update `Nal::len`, `Nal::next_piece_idx`, `Nal::pieces` & create new `Nal`s on boundary
end_rtp_nal => umm whadoido? append_rtp_nal would've already updated the `Nal::len` and `Nal::next_piece_idx` of final `Nal` correctly
```
What would be the purpose of `end_rtp_nal`? How would we represent `ending current NAL`? Maybe we could use it to reset the states (`did_find_boundary`, `seen_one_zero` etc)?

[the remark this answers is not in the document]
Comment (scottlamb) [reply src/codec/h264.rs]: Good point; now that you mention it, I'm not sure we need `end_rtp_nal`.

I do want to make sure we handle the case where an RTP packet ends with a zero. Until we get the next packet, we don't know for sure if that's part of a NAL or if it's part of an Annex B boundary. If `append_rtp_nal` sort of optimistically assumes it's part of a NAL and then the following one walks that back if it finds the rest of the boundary, then maybe `end_rtp_nal` isn't necessary (and resetting state can happen at the beginning of the next `start_rtp_nal`). If `append_rtp_nal` works a different way, maybe there's something for `end_rtp_nal` to do.

[the remark this answers is not in the document]
Comment (thatdevsherry) [reply src/codec/h264.rs]: Yup, I'm steering towards letting `append_rtp_nal` handle boundaries b/w packets since it's already doing all the splitting work.

Judging by the time `end_rtp_nal` would be called, it would be doing anything but reading the RTP payload.

However, will try thinking if there's a way to use `end_rtp_nal`. Not having `end` and just `start` & `append` feels off :p

[the remark this answers is not in the document]
Comment (thatdevsherry) [reply src/codec/h264.rs]: What if we instead of having separate `start_rtp_nal` & `append_rtp_nal`, we have a single `NalParser::push(rtp_payload: Bytes)` that handles processing the payload into `NalParser::nals` and `NalParser::pieces`?

It would act as a single API that internally manages `nals` and `pieces`, and would also handle splitting Annex B streams (and edge case if boundaries span b/w packets).

What do you think?

File: src/codec/h264.rs [hunk 6]
```diff
@@ -36,13 +36,182 @@ pub(crate) struct Depacketizer {
...
 
     /// In state `PreMark`, an entry for each NAL.
     /// Kept around (empty) in other states to re-use the backing allocation.
     nals: Vec<Nal>,
+
+    // If true, retina does not panic when the header of a FU-A frag does not match
+    // the header of the last NAL saved.
+    ignore_inconsistent_fu_a_headers: bool,
+
+    // annex b delimiter watcher state
```
[the remark this answers is not in the document]
Comment (thatdevsherry) [reply src/codec/h264.rs]: Addressed in [bee29b9](https://github.com/scottlamb/retina/pull/71/commits/bee29b99357eb1104ce923f78d9f9bdee56b0b51).

Since the zero state is currently being used inside a single function, I've removed it from the struct and moved the state inside the function body.

File: src/codec/h264.rs (hunk 1 above)
Comment (thatdevsherry) [reply src/codec/h264.rs:1819]: Here's how I understand this:

```rust
let rtp_payloads: Vec<Bytes> = vec!(b"previous-nal\x00\x00", b"\x00new-nal"); // code isn't valid but just for understanding

w/o handling boundaries b/w packets, we'd get:

pieces: [ [b"previous-nal\x00\x00"], [b"\x00new-nal"] ]

what we should be forming:

pieces: [ [b"previous-nal"], [b"new-nal"] ]
```

To handle this, we can do something like:

```rust
append_rtp_nal(...) {
  // if first index is 0x00, check if last `piece` ends with two 0x00. if so, remove them and also this pkt's 1st index i.e. 0x00
  // if first two indexes are 0x00, again check if last `piece` ends with 0x00. If so, remove them and this pkt's first two indexes
}
```

However, this would fail for **byte-by-byte**, because, as per my understanding, `pieces` would end up like this when we reach the third boundary:

```rust
pieces: [ .., [b"n"], [b"a"], [b"l"], [b"\x00], [b"\x00]] // each append_rtp_call adds the given data to pieces
```

If we look at it from real-world usage, there shouldn't be a case where three `0x00` of a boundary are split b/w three `pieces`, rather the possible cases can be:

```rust
// first `0x00` of boundary ends and remaining two `0x00` are at the start of next packet
pkts = [ [b"...\x00"], [b"\x00\x00.."] ]

// or first two `0x00` of boundary end and remaining `0x00` is at the start of next packet
pkts = [ [b"..\x00\x00"], [b"\x00..] ]
```

Is there a gap in my understanding? I can't figure out what would **byte-to-byte** test give us.

If what I understand is right, and we want to test **byte-to-byte**, then maybe we can bring back `end_rtp_nal` and get a flow like this:

```
start_rtp_nal: creates a new Nal as discussed
append_rtp_nal: will not store directly into `pieces`, but rather in an intermediary state where it can handle boundaries byte-by-byte
end_rtp_nal: move over data from intermediary state into actual `pieces`
```

Please let me know if I'm missing something.

File: src/codec/h264.rs [hunk 7]
```diff
@@ -36,13 +36,182 @@ pub(crate) struct Depacketizer {
...
+            pieces: Vec::new(),
+            nals: Vec::new(),
+            ignore_inconsistent_fu_a_headers: false,
+            seen_one_zero_at: None,
+            seen_second_zero_at: None,
+            did_find_boundary: false,
+        }
+    }
+
+    /// Helper function to clear `NalParser::nals` and `NalParser::pieces` vectors.
```
[the remark this answers is not in the document]
Comment (thatdevsherry) [reply src/codec/h264.rs:72]: Indirectly addressed in [bee29b9](https://github.com/scottlamb/retina/pull/71/commits/bee29b99357eb1104ce923f78d9f9bdee56b0b51) and [41c2c8f](https://github.com/scottlamb/retina/pull/71/commits/41c2c8fe3512dfde6d5502be9e25c23bef0140a5).

Since state currently isn't being maintained outside of `break_apart_nals`, there's nothing to cleanup.

**Comment by thatdevsherry** [comment]
Should `NalParser` ignore continuous zeros? Currently `benches/client.rs` is failing because it creates a long `0x00` bytes data, which `break_apart_nals` fails to parse because it expects that three consecutive `0x00` are a boundary, and not part of a nal.

`benches/client.rs` is sending a FU-A (start) with all bytes being `0x00`. Isn't this test data incorrect?

**Comment by scottlamb** [comment]
Apologies—I've been quite bogged down. Will look through your comments and updates tomorrow.

**Comment by thatdevsherry** [comment]
> Apologies—I've been quite bogged down. Will look through your comments and updates tomorrow.

Wishing you good health. Please take your time to recover.

File: src/codec/h264.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (scottlamb) [reply src/codec/h264.rs]: Huh, I'm confused. The view I'm seeing in the GitHub review isn't reflecting the changes we discussed in this thread. It shows `in_fu_a` as a bool and has this `ignore_inconsistent_fu_a_headers` stuff. Did you revert these changes?

It's also possible I'm just having trouble with the GitHub review interface—I confess I still have trouble navigating it, sometimes send off a bunch of separate single comments because I can't find the "Start a review" button to batch them in some flows, etc.

I want to be sure I'm looking at the right thing when commenting to avoid causing more confusion...

**Review by scottlamb — COMMENTED** [review]
Apologies again for the long delay—much longer than I expected!

> Should NalParser ignore continuous zeros? Currently benches/client.rs is failing because it creates a long 0x00 bytes data, which break_apart_nals fails to parse because it expects that three consecutive 0x00 are a boundary, and not part of a nal.

Great question. I looked back at the spec and realized something I don't think I was clear about (may have forgotten about entirely). The separator can have extra leading zeros, so if the bytestream has `0x12 0x34 0x00 0x00 0x00 0x00 0x1 0x56 0x78`, then the NALs are supposed to be just `0x12 0x34` then just `0x56 0x78`. So basically yes, they should be ignored.

The `benches/client.rs` data is definitely not a valid, just some easy dummy value that worked with the current implementaiton. Feel free to change it to say a long string of `0x04`s or whatever instead to avoid tripping on the new logic.

File: src/codec/h264.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (scottlamb) [reply src/codec/h264.rs]: > Huh, I'm confused.

Never mind. I think I made the mistake of clicking through an "resolved conversation", and it then showed me an older state of the review. If I don't do that, I see the current state.

File: src/codec/h264.rs [hunk 8]
```diff
@@ -45,6 +61,170 @@ pub(crate) struct Depacketizer {
...
+    /// to break apart NALs from the Annex B stream.
+    fn break_apart_nals(&mut self, data: Bytes) -> Result<bool, String> {
+        // Should `&mut Bytes` be passed instead of doing this?
+        let mut data_copy = data;
+
+        let mut nal_start_idx = 0;
+        let mut did_find_boundary = false;
+
+        // Check if current packet starts with either a single or two 0x00 bytes.
+        // If so, check if previous piece ended with either one or two 0x00 bytes.
```
Comment (scottlamb) [inline src/codec/h264.rs:96]: fwiw, it might be simpler to keep in `NalParser` the current count of trailing zeros. Then we know how far back we need to trim if that's necessary, and we can also handle the case in which the two zeros were in previous calls and the `0x03` is the starting byte of the current call (your TODO on line 143).

Comment (thatdevsherry) [reply src/codec/h264.rs:96]: Thanks for the suggestion.

If we keep the count of zeros inside `NalParser`, could something like this make sense?

```
pkt_1 = [..., 0x00, 0x00]

# count will get incremented to 2

pkt_2 = [0x00, 0x00, 0x00, ...] # if we have more than 3 zeros, we will remove all

# count will get incremented to 5
```

We can then write a function that would run when we read the first non-boundary character, which would then backtrack and remove zeros, going backwards across packets until the count gets decremented to `0`.

```
current_pkt = [ 0x00, 0x00, 0x00, ... ]
pieces = [ [ ..., 0x00, 0x00 ] ]

# we had read the previous pkt, so it is in `pieces`, current one is still being read so it won't be in `pieces`

# going backwards from current pkt, we'll remove the three zeros, and get

current_pkt = [ ... ]
pieces = [ [ ..., 0x00, 0x00 ] ]
count = 2

# since we've reached the start of current pkt and count isn't 0, we'll move over to the last pkt in `pieces`, and continue decrementing from end.

current_pkt = [ ... ]
pieces = [ [ ... ] ]

count = 0
```

The zero removing function can do something like this.

```
1. get the first non-boundary character index inside our current pkt, index is `z`
2. create a reversed array from current_pkt[0..z] and loop over it, let's call this rev_arr_processed
3. pop each value and decrement count
4. if count reaches 0, end by joining `[ rev_arr_processed.rev() .. current_pkt[z..] ]`, else
5. if iteration ends and count is not 0, get the last pkt from `pieces`
6. repeat from step 2, except that when this ends, we have to update the pkt(s) in `pieces`
```

The algorithm can work for a single pkt as well as across pkts, and will not limit itself to just 3 zero boundaries. If we have a few zeros inside a single pkt:

```
## Single pkt flow

current_pkt = [ a..., 0x00, 0x00, 0x00, z... ]

# `a...` are non-boundary bytes before boundary, and `z...` are non-boundary bytes after the boundary

# processing overview

rev_arr_processed = [ 0x00, 0x00, 0x00, ...a ]
# after processing
rev_arr_processed = [ ...a ]

# join them back
[ rev_arr_processed.rev() .. current_pkt[z..] ]

result = [ a..., z... ]
```

Please let me know if my explanation isn't easy to understand, I'll improve it.

Also, please excuse if my understanding has gotten rusty (pun absolutely intended), came back to this project after a while 😄

Comment (thatdevsherry) [reply src/codec/h264.rs:96]: I've created a [prototype in python for the above mentioned algorithm](https://github.com/thatdevsherry/retina-packet-parser-poc). Thought it's easier to write a PoC first instead of directly coding in Rust.

If I can handle the cases in the PoC, I'll move over the logic to retina and update the PR.

Comment (scottlamb) [reply src/codec/h264.rs]: Awesome. I tried adding a test that has the boundary spread across multiple packets. Would you expect this to pass? It doesn't seem to (although I might not be understanding your Python code's contract correctly).

```python
    def test_boundary_spread_across_pkts(self):
        input = [
            [1, 2, 3, 4, 0],
            [0],
            [0],
            [1],
            [5, 6, 7, 8],
        ]
        nal_parser = NalParser()
        for idx, i in enumerate(input):
            nal_parser.process_packet(i)
        output = nal_parser.pieces
        expected = [
            [1, 2, 3, 4],
            [5, 6, 7, 8],
        ]
        self.assertEqual(output, expected)
```

Comment (thatdevsherry) [reply src/codec/h264.rs]: The PoC currently defines the boundary as **consecutive `0x00`s >= 3**.

The above logic would make it that our test should return something like `[ [1,2,3,4], [1], [5,6,7,8] ]` since it isn't treating `[0, 0, 0, 1]` as a boundary. Would `[0, 0, 0, .. , 1]` be a valid boundary? Could you also let me know what other variations we should count as boundaries?

The PoC is pretty dumb and finds a single pkt, single byte `[0]` weird. The output of our test is:

```
[[1, 2, 3, 4], None, None, [1], [5, 6, 7, 8]]
```

which is near what I expected at the start.

I guess we can use the following constraint to properly define the full boundary.

- consecutive `0x00`s >= 2, followed by a number in `[0, 1, 2, 3]`

Does the above case cover all boundaries? You also mentioned in notion that a delimiter is `0,0,1`, and that `3` is used as an escape e.g. `0,0,1 -> 0,0,3,1`. So I'm not sure if above constraint would be valid.

**Comment by thatdevsherry** [comment]
Closing in favor of #100
